tokio-net/src/driver/reactor/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,299 @@
+
+    fn as_usize(&self) -> usize {
+        self.addr
+    }
+
+    fn from_usize(addr: usize) -> Self {
+        debug_assert!(addr <= Self::BITS);
+        Self { addr }
+    }
+}
+
+pub(in crate::driver) type Iter<'a> =
+    std::iter::FilterMap<std::slice::Iter<'a, Slot>, fn(&'a Slot) -> Option<&'a ScheduledIo>>;
+
+pub(crate) struct Local {
+    head: CausalCell<usize>,
+}
+
+pub(crate) struct Shared {
+    remote_head: AtomicUsize,
+    size: usize,
+    prev_sz: usize,
+    slab: CausalCell<Option<Box<[Slot]>>>,
+}
+
+impl Local {
+    pub(crate) fn new() -> Self {
+        Self {
+            head: CausalCell::new(0),
+        }
+    }
+
+    #[inline(always)]
+    fn head(&self) -> usize {
+        self.head.with(|head| unsafe { *head })
+    }
+
+    #[inline(always)]
+    fn set_head(&self, new_head: usize) {
+        self.head.with_mut(|head| unsafe {
+            *head = new_head;
+        })
+    }
+}
+
+impl Shared {
+    const NULL: usize = Addr::NULL;
+
+    pub(crate) fn new(size: usize, prev_sz: usize) -> Self {
+        Self {
+            prev_sz,
+            size,
+            remote_head: AtomicUsize::new(Self::NULL),
+            slab: CausalCell::new(None),
+        }
+    }
+
+    #[cold]
+    fn fill(&self) {
+        #[cfg(test)]
+        test_println!("-> alloc new page ({})", self.size);
+
+        debug_assert!(self.slab.with(|s| unsafe { (*s).is_none() }));
+
+        let mut slab = Vec::with_capacity(self.size);
+        slab.extend((1..self.size).map(Slot::new));
+        slab.push(Slot::new(Self::NULL));
+        self.slab.with_mut(|s| {
+            // this mut access is safe — it only occurs to initially
+            // allocate the page, which only happens on this thread; if the
+            // page has not yet been allocated, other threads will not try
+            // to access it yet.
+            unsafe {
+                *s = Some(slab.into_boxed_slice());
+            }
+        });
+    }
+
+    #[inline]
+    pub(crate) fn alloc(&self, local: &Local) -> Option<usize> {
+        let head = local.head();
+        #[cfg(test)]
+        test_println!("-> local head {:?}", head);
+
+        // are there any items on the local free list? (fast path)
+        let head = if head < self.size {
+            head
+        } else {
+            // if the local free list is empty, pop all the items on the remote
+            // free list onto the local free list.
+            self.remote_head.swap(Self::NULL, Ordering::Acquire)
+        };
+
+        // if the head is still null, both the local and remote free lists are
+        // empty --- we can't fit any more items on this page.
+        if head == Self::NULL {
+            #[cfg(test)]
+            test_println!("-> NULL! {:?}", head);
+            return None;
+        }
+
+        // do we need to allocate storage for this page?
+        let page_exists = self.slab.with(|s| unsafe { (*s).is_none() });
+        if page_exists {
+            self.fill();
```

Comment:
The `page_exists` name confused me. Maybe name it `page_needs_alloc` or something.